handlers.go

```diff
@@ -111,158 +179,594 @@ func handleMachines(add func(Change), machines map[string]*charm.MachineSpec, de
+	if p.newUnitsPlacement == nil {
+		p.newUnitsPlacement = make(map[string][]*AddUnitChange)
+	}
+	key := appName + "/" + placement.Application
+	changes := p.newUnitsPlacement[key]
+	if changes == nil {
+		newUnits := p.appChanges[placement.Application]
+		if newUnits == nil {
+			changes = []*AddUnitChange{}
 		} else {
-			number = 0
+			// Copy the slice for our purposes as we are going to consume the
+			// resulting slice.
+			changes = newUnits[:]
 		}
-		servicePlacedUnits[placement.Application] = number
+		p.newUnitsPlacement[key] = changes
 	}
-	otherUnit := fmt.Sprintf("%s/%d", placement.Application, number)
-	parentId = records[otherUnit].Id()
-	if placement.ContainerType != "" {
-		parentId = addContainer(add, placement.ContainerType, parentId, series, appConstraints)
+	if len(changes) == 0 {
+		return nil
+	}
+	result, changes := changes[0], changes[1:]
+	p.newUnitsPlacement[key] = changes
+	return result
+}
+
+func (p *unitProcessor) definedApplicationForUnit(appName string, application *charm.ApplicationSpec, placement *charm.UnitPlacement, directive string) unitPlacement {
+	appendPlacement := func(result unitPlacement) unitPlacement {
+		result.placementDescription += " to satisfy [" + directive + "]"
+		return result
+	}
+	// First see if there is a unit of the placement application that doesn't
+	// have a unit of the application we are trying to place next to it (or in
+	// a container as defined by the placement).
+	existingMachine := p.nextMachineForExistingAppUnits(appName, placement)
+	if existingMachine != "" {
+		return appendPlacement(p.machinePlacement(existingMachine, placement.ContainerType))
+	}
+	// If there are none in the model, look for units of appName that have been placed.
+	change := p.nextUnitChangeForApp(appName, placement)
+	if change != nil {
+		var target string
+		placeholderMachine := isPlaceholder(change.Params.baseMachine)
+		placeholderDescription := " machine " + change.Params.baseMachine
+		if placeholderMachine {
+			placeholderDescription = "new" + placeholderDescription
+			target = change.Id()
+		} else {
+			placeholderDescription = "existing" + placeholderDescription
+			target = change.Params.baseMachine
+		}
+		result := unitPlacement{
+			target:               target,
+			placeholder:          placeholderMachine,
+			placementDescription: placeholderDescription,
+			baseMachine:          change.Params.baseMachine,
+		}
+		// It would be nice if we could be smarter with the creation of containers.
+		// Need to check with the GUI folks about removing container additions, and
+		// instead just handling it in unit placement.
+		if placement.ContainerType != "" {
+			result = p.addContainer(result, application, placement.ContainerType)
+		}
+		return appendPlacement(result)
+	}
+
+	return unitPlacement{
+		baseMachine: placeholder(p.getNextMachinePlaceholderID()),
+	}
+}
+
+type unitPlacement struct {
+	target               string
```

Comment:
Can you give some explanation and examples of what target contains here? Sometimes it's a change id and sometimes it's a machine, and I don't really understand what that means.